Add initial OpenDev ARM64 testing

ianw opened the pull request:
This is the initial configuration for Zuul to run pyca/cryptography
tox jobs on a range of ARM64 nodes provided by OpenDev.  The
underlying ARM64 resources are donated for use by the OpenDev project
by Linaro.

This is under discussion at https://github.com/pyca/cryptography/issues/5339

If the OpenDev Zuul app (https://github.com/apps/opendev-zuul) is
added to this repository, it should be able to speculatively test and
run these jobs (however, some configuration will be required on the
OpenDev side before this will happen).

This is currently a very simple run of tox on the code.  For basic job
documentation see https://zuul-ci.org/docs/zuul/reference/jobs.html.

These jobs inherit from the opendev base job defined in
https://opendev.org/opendev/base-jobs.  This handles the node setup,
initial clone of pull requests, etc. and then after the job runs the
log collection, upload and publishing steps.  This in turn uses a lot
of reusable components from https://zuul-ci.org/docs/zuul-jobs/

ianw commented:
Note there is no need to merge this at this time, it is just an initial way to show what the jobs will look like

ianw commented:
recheck

ianw commented:
recheck

ianw commented:
recheck

ianw commented:
recheck

ianw commented:
recheck

ianw commented:
recheck

ianw commented:
This is now ready for review.  As you can see Zuul is driven by a Ansible playbook that means we can do things in Ansible, or shell out, or use any of our prebuilt bits and pieces from https://zuul-ci.org/docs/zuul-jobs/

alex commented on playbooks/main.yaml:
```diff
@@ -0,0 +1,32 @@
+- hosts: all
```
Is there any chance this file could go into `.zuul.d/`?

ianw replied:
Currently Zuul considers anything in ```.zuul.d``` configuration, so putting playbooks in here will result in configuration errors.

Two thoughts are that we could move it to a sibling hidden-directory, or we could enhance Zuul to obey a ```.zuul.ignore``` type file in directories so it knows they aren't config.  I'll try the first and investigate the second.

alex replied:
`.zuul-playbooks/` is fine. just don't want to have a top level playbooks dir.

ianw replied:
> `.zuul-playbooks/` is fine. just don't want to have a top level playbooks dir.

Done like this for now

reaperhulk commented:
Is it possible for zuul to put the content of those comments into the github status updates? It's unfortunate that it provides good links in the comment but a largely worthless details link with a consolidated pass/fail as a status update.

ianw commented:
> In terms of how this works, it'll auto run on all PRs I assume? 

Yes; to be concrete the configuration is at https://opendev.org/pyca/project-config/src/branch/master/zuul.d/pipelines.yaml#L31 where it is set up to listen to pull request events, or somebody adding a comment of "recheck" on a pull-request.

> What about commits to master?

I believe we could listen to "push" requests.  

https://zuul-ci.org/docs/zuul/reference/drivers/github.html#value-pipeline.trigger.%3Cgithub%20source%3E.event.push

This wouldn't have anywhere to report the status, however?  You would have to know to go and look at something like the build results page https://zuul.opendev.org/t/pyca/builds to see the results of the run.  If you'd like this we can investigate it.

> Is it possible for zuul to put the content of those comments into the github status updates? It's unfortunate that it provides good links in the comment but a largely worthless details link with a consolidated pass/fail as a status update.

I will have to ask for advice on that.  I think @felixedel might have some helpful insight here.  I think there are some limitations around the check status API (https://zuul-ci.org/docs/zuul/discussion/github-checks-api.html#design-decisions) and reporting collections of jobs.

alex commented:
Non-PR commits also get statuses, you can see them on https://github.com/pyca/cryptography/commits/master

ianw commented:
recheck

ianw commented:
> Is it possible for zuul to put the content of those comments into the github status updates?

After investigation we are currently reporting via the "status API", which we did find was not setting a useful URL (fixed with https://review.opendev.org/#/c/744570/).

However, we can do better by using the "checks API" so that the results show up in the "checks" results tab and the links will all be more useful too.  As you can see I've switched the config to use this but we got

> Failed to update check run pyca/check: 403 Resource not accessible by integration

Upon investigation, our bot didn't request "checks" access, so it can't use this API.  I've updated the bot to request this permission now; @alex as you originally accepted the bot I think you'll need to re-approve it with the new checks permission.  This should allow the better reporting.

alex commented:
Re-approved already!

On Mon, Aug 3, 2020 at 7:41 PM Ian Wienand <notifications@github.com> wrote:

> Is it possible for zuul to put the content of those comments into the
> github status updates?
>
> After investigation we are currently reporting via the "status API", which
> we did find was not setting a useful URL (fixed with
> https://review.opendev.org/#/c/744570/).
>
> However, we can do better by using the "checks API" so that the results
> show up in the "checks" results tab and the links will all be more useful
> too. As you can see I've switched the config to use this but we got
>
> Failed to update check run pyca/check: 403 Resource not accessible by
> integration
>
> Upon investigation, our bot didn't request "checks" access, so it can't
> use this API. I've updated the bot to request this permission now; @alex
> <https://github.com/alex> as you originally accepted the bot I think
> you'll need to re-approve it with the new checks permission. This should
> allow the better reporting.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pyca/cryptography/pull/5341#issuecomment-668292883>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAGBGLRIQCIHQ6EVVXFULR65DJXANCNFSM4PFJLJVQ>
> .
>


-- 
All that is necessary for evil to succeed is for good people to do nothing.

ianw commented:
recheck

ianw commented:
Ok this is reporting via the checks api now.  This change will have an old entry left over from the prior updates with the status API, but from now on they'll be in the checks tab and the link will be like https://github.com/pyca/cryptography/pull/5341/checks?check_run_id=942805616

ianw commented:
To summarise prior comments

* links in reports : switched to checks API so consistent with other CI results
* config/playbooks : moved to separate directory.  I have also proposed a method for Zuul to ignore non-config files in https://review.opendev.org/744811 which would allow a single-directory model in the future.
* reports on non-PR master merges : this should be possible by configuring Zuul to watch for a ``push`` event in  https://opendev.org/pyca/project-config/src/branch/master/zuul.d/pipelines.yaml#L31.  TBH we don't usually run CI on merged changes because Zuul does the merge and only after ensuring CI passes (it's kind of it's value proposition :).  Nothing will happen till this config merges, so my preference would probably be to merge PR support like this, and then switch on ``push`` tracking; leaving us to tackle any issues post-merge testing throws separately (which is theoretically none, but ... first time and all).

I'm happy to address further comments :)
